File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
import math
def solution(numbers):
    
    def check(n): #소수 확인 함수
        if n==0 or n==1:
            return False
        for i in range(2, int(n**0.5)+1):
            if n%i==0:
                return False #소수가 아님
        return True
    
    arr=[]
    for i in range(1, len(numbers)+1):
        for n in permutations(numbers, i):
            arr.append(int(''.join(n)))
    
    arr=list(set(arr))
    answer = 0
    for j in arr:
        if check(j):
            answer+=1
            print(j)
        else:
            continue
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    static char[] d;
    static boolean[] used;
    static HashSet<Integer> made;   // 중복 숫자 제거용

    public int solution(String numbers) {
        d = numbers.toCharArray();
        used = new boolean[d.length];
        made = new HashSet<>();

        dfs("");                    // 길이 0에서 시작, 길이 1~len까지 전부 만들기

        int ans = 0;
        for (int x : made) if (isPrime(x)) ans++;
        return ans;
    }

    static void dfs(String cur) {
        if (!cur.isEmpty()) {
            made.add(Integer.parseInt(cur));      // 앞의 0은 parseInt가 알아서 처리
        }
        if (cur.length() == d.length) return;

        for (int i = 0; i < d.length; i++) {
            if (used[i]) continue;
            used[i] = true;
            dfs(cur + d[i]);                      // 현재 자리 하나 붙이기
            used[i] = false;
        }
    }

    static boolean isPrime(int n) {
        if (n < 2) return false;                  // 0, 1 제외
        int r = (int)Math.sqrt(n);
        for (int i = 2; i <= r; i++) if (n % i == 0) return false;
        return true;
    }
}
